gax/src/test/java/com/google/api/gax/batching/v2/BatcherImplTest.java

```diff
@@ -268,4 +445,53 @@ private void testElementTriggers(BatchingSettings settings) throws Exception {
     assertThat(result.get()).isEqualTo(16);
     assertThat(anotherResult.isDone()).isTrue();
   }
+
+  private class ElementAddThread extends Thread {
+    private volatile boolean isCancelled;
+    private final AtomicInteger counter = new AtomicInteger();
+    private final Batcher<Integer, Integer> batcher;
+
+    ElementAddThread(Batcher<Integer, Integer> batcher) {
+      this.batcher = batcher;
+    }
+
+    @Override
+    public void run() {
+      while (!isCancelled) {
```

Comment:
Why not make it deterministic and have the number of elements be preset and then use a settable future for the test to block on

Reply:
If we block on the future then the tests take **~1500ms** to finish just for 1K elements, considering `Thread.sleep(1)` after each addition and 3K times `flush()` (which is also blocking). 

```java
  public void run() {
    List<ApiFuture<Integer>> results = new ArrayList<>();
    try {
      for (int i = 0; i < 1000; i++) {
        results.add(batcherTest.add(i));
        Thread.sleep(1);
      }
      ApiFutures.allAsList(results).get();
      assertThat(results).hasSize(1000);
    } catch (Exception ex) {
      throw new AssertionError(ex);
    }
  }
```

1K elements does recreates the duplicate element scenario if `currentOpenBatch` is **not atomically swapped**(less than that element can pass the test without conflict).

Reply:
I guess leave it as is